src/bpftrace.cpp

```diff
@@ -1308,10 +1331,75 @@ int BPFtrace::setup_perf_events()
   return 0;
 }
 
-void BPFtrace::poll_perf_events(bool drain)
+int BPFtrace::setup_ringbuf()
+{
+  ringbuf_ = static_cast<struct ring_buffer *>(bpf_new_ringbuf(
+      maps[MapManager::Type::Ringbuf].value()->mapfd_, ringbuf_printer, this));
+  // init ringbuf loss counter
+  if (bpf_update_elem(
+          maps[MapManager::Type::RingbufLossCounter].value()->mapfd_,
+          &rb_loss_cnt_key_,
+          &rb_loss_cnt_val_,
+          0))
+  {
+    LOG(ERROR) << "fail to init ringbuf loss counter";
+    return -1;
+  }
+  return 0;
+}
+
+void BPFtrace::teardown_output()
 {
-  const int timeout_ms = 100;
+  if (feature_->has_map_ringbuf())
+  {
+    bpf_free_ringbuf(ringbuf_);
+    // print ringbuf event losses
+    uint64_t value = 0;
+    if (bpf_lookup_elem(
+            maps[MapManager::Type::RingbufLossCounter].value()->mapfd_,
+            &rb_loss_cnt_key_,
+            &value))
+    {
+      LOG(ERROR) << "fail to get ringbuf loss counter";
+    }
+    if (value)
+    {
+      out_->lost_events(value);
+    }
```

Comment:
Looks like this will only print at the end of the trace session. Perhaps fine, but I wonder if we should be sampling the loss counter every N events or something. So that lost events are displayed to the user during the trace session (rather than at the end).